### ISSUE TITLE
sdcTable::protectLinkedTables can be unsafe

### ISSUE THREAD

**Issue by olangsrud** [sdcTools/UserSupport#136]
**SDC tool used:** sdcTable
**Version used:**  0.31
**Operating system used:**  Windows

---
Below is small example that demonstrates that suppressing linked tables by iteration can be unsafe. This issue has become relevant since I've made available a method to replace suppressed cells with decimal numbers. When the suppression is not safe, as below, such decimal numbers cannot be made (see easySdcTable::PTxyz).
```
library(sdcTable)

# Data including two aggregating variables.
z <- data.frame(region = c("A", "B", "I", "J", "K", "L", "M", "N"), 
                county = c(rep("county-1", 4), rep("county-2", 4)), 
                group = c(rep("small", 5), "BIG", "BIG", "small"), 
                n = c(2, 2, 8, 8, 3, 3, 3, 7), 
                stringsAsFactors = FALSE)


# Defining hierarchies
dimList1 <- list(region = data.frame(levels = c("@","@@","@@@","@@@","@@@","@@@","@@","@@@","@@@","@@@", "@@@"), 
                                     codes = c("Total","county-1","A","B","I","J","county-2","K","L","M","N"), 
                                     stringsAsFactors = FALSE))
dimList2 <- list(region = data.frame(levels = c("@","@@","@@@","@@@","@@","@@@","@@@","@@@","@@@", "@@@", "@@@"), 
                                     codes = c("Total","BIG","L","M","small","A","B","I","J","K","N"), 
                                     stringsAsFactors = FALSE))

# Defining common cells
commonCells <- list(region = list("region", c("Total", "A", "B", "I", "J", "K", "L", "M", "N"))[c(1, 1, 2, 2)])


# makeProblem and primarySuppression in sdcTable 
problem1 <- makeProblem(z, dimList1, dimVarInd = 1, freqVarInd = 4)
problem2 <- makeProblem(z, dimList2, dimVarInd = 1, freqVarInd = 4)
primary1 <- primarySuppression(problem1, type = "freq", maxN = 3)
primary2 <- primarySuppression(problem2, type = "freq", maxN = 3)

# Secondary suppression of linked tables 
protected <- protectLinkedTables(primary1, primary2, commonCells, method = "SIMPLEHEURISTIC")

# Look at output
getInfo(protected[[1]], type = "finalData")
getInfo(protected[[2]], type = "finalData")


#  The result is that no secondary suppression is needed. Although the tables are safe 
#  individually, suppressed cells can be revealed by looking at both. The difference between 
#  "county-1" and "small" is two cells, K and N. The difference is 10 and N is 7. 
#  Thus B can be calculated as 10-7 = 3.
```

**Comment by bernhard-da** [comment]
@olangsrud thx for reporting. i can reproduce and will try to fix.

**Comment by olangsrud** [comment]
In the example above, a suppressed cell could be computed from three non-suppressed cell, 30-20-7 = 3. The general problem is whether there is any linear combination of non-suppressed cells that can be used to calculate the value of any suppressed cell.

**Comment by bernhard-da** [comment]
hi, i just pushed a possible fix in https://github.com/sdcTools/sdcTable/commit/07c7baf01867a8bc7876c7717ce086ab2ad3a5a3; could you please test with 0.31.1 and tell me, if it works as expected?

**Comment by olangsrud** [comment]
@bernhard-da  Is it possible for you to make a windows binaries zip installation file (again)?

**Comment by bernhard-da** [comment]
@olangsrud yes, you can download a binary for windows [**here**](https://cloud.bmeindl.eu/f/cbe50b1aedbc4bf5aba7/?dl=1) (link is valid for until     2020-05-02 09:56:49)

**Comment by olangsrud** [comment]
Thanks, I have downloaded the file.

**Comment by olangsrud** [comment]
I have tested the new version. The problem in the example above is solved. It seems that some problems are solved, but the general problem I mentioned above (…any linear combination…) is not solved. This time I haven’t produced an easy interpretable sdcTable example. Instead I refer to results using the interface from easySdcTable.

```
library(easySdcTable)
z2 <- EasyData("z2")
ProtectTable (z2, 1:4, 5, method="SIMPLEHEURISTIC")$data
```

With the new version, as in version sdcTable 0.31, two suppressed cells can be revealed. The two cells has values 2 and 36 and this can be seen in the example of  `easySdcTable::PTxyz`.

Another and more complicated example is 

```
z3 <- EasyData("z3")
ProtectTable(z3, 1:6, 7, method="SIMPLEHEURISTIC",  protectZeros = FALSE)$data
```

Using version 0.31 of sdcTable, 12 cells could be revealed. With the new version, 5 cells could be revealed.  This example also demonstrates that the new version is much more time consuming. My laptop (relatively old) spent 3 seconds using version 0.31 and 97 seconds with the new version. 
An even bigger example is:

```
z4 <- EasyData("sosialFiktiv")
ProtectTable(z4, 1:6, 7, method="SIMPLEHEURISTIC", maxN = 2, protectZeros = FALSE)$data
```
Using version 0.31 of sdcTable, 8 cells could be revealed. My laptop spent 9 seconds using version 0.31. I tried the new version, but after one hour of computing time, I stopped it.

**Comment by olangsrud** [comment]
I have unassigned me, but I do not have privileges to remove the awaiting_feedback label myself.

**Comment by bernhard-da** [comment]
@olangsrud thx. in the meantime, i have made some progress; given your examples, i was able to identify the root of the problem; however i am not sure how to (generally) solve this. if i find some time in the next few weeks, I might work on this further or just remove the functionality as I am personally not using it anyways.

**Comment by bernhard-da** [comment]
hi @olangsrud: i finally had some time to tacke this; i have rewritten the entire simpleheuristic-approach so that it does not follow the argus (hitas) approach of travelling subtables any longer but to focus directly on linear constraints; for the linked-table approach, i now additionally generate the required constraints that would lead to the differencing attack you described; eg when two cells in the two table differ exactly by two common cells; 

i would be very happy if you could this before I draft a new release based on this; a zip-file for windows can be found [**here**](https://win-builder.r-project.org/9vCu0L0uArfP/sdcTable_0.31.99.zip)

**Comment by olangsrud** [comment]
Hi @bernhard-da: Yes, I will do some checking concerning revealing supressed cells by the method I have used above.

**Comment by bernhard-da** [comment]
> Hi @bernhard-da: Yes, I will do some checking concerning revealing supressed cells by the method I have used above.

:+1: thx!

**Comment by olangsrud** [comment]
Note first that I have changed default method within easySdcTable to a method called "SimpleSingle" which is "SIMPLEHEURISTIC" with detectSingletons=TRUE  or with threshold=1 depending on allowZeros. Now with sdcTable version 0.31.99, the bottom example of  easySdcTable::PTxyz reveals no cells. But if this is changed from "SimpleSingle" to "SIMPLEHEURISTIC" two cells can still be revealed. The code to protect data is:
    
    library(easySdcTable)
    z2 <- EasyData("z2")
    ProtectTable (z2, 1:4, 5, method="SIMPLEHEURISTIC")$data

To explain what’s going on, the rows of the input data within hovedint=="trygd" (hovedint is a dimension) is below together with suppression results:  

    region hovedint fylke kostragr freq sdcStatus suppressed
         A    trygd     1      300   36         x         NA
         B    trygd     4      300   18         s         18
         C    trygd     5      300   25         s         25
         D    trygd     5      300   13         s         13
         E    trygd     6      300   52         s         52
         F    trygd     6      300   22         s         22
         G    trygd     8      300    8         x         NA
         H    trygd     8      300   15         x         NA
         I    trygd     1      400    2         u         NA
         J    trygd    10      400   20         x         NA
         K    trygd    10      400   11         x         NA

Aggregates within "trygd" are not suppressed.  Especially the 400-sum (=33) within kostragr and the 10-sum (=31) within fylke is safe. The difference (=2) is I-trygd, a suppressed value. 

In my bigger example datasets, I haven’t seen cases where it is possible to reveal cells from linked-table-suppression using sdcTable version 0.31.99. But a concern is computing time, the number of supressed cells and errors (we cannot find a solution; no cells left to suppress).   An example causing error is
 
    sosialFiktiv = EasyData("sosialFiktiv")
    ProtectTable(sosialFiktiv, 1:6, 7, method= "SimpleSingle",  protectZeros = FALSE)

If protectZeros (= allowZeros) is changed to TRUE, the number of suppressed cells is 12644 and the computing time is 9 minutes on my old laptop.  

As mention to earlier I have made a function in the package SSBtools called GaussSuppression (Secondary suppression by Gaussian elimination). If you can create the x-matrix needed, it would also be possible to make use of it within sdcTable as an alternative suppression method. I have already implemented it as an alternative method within easySdcTable.  In the above example (protectZeros=TRUE), the number of suppressed cells is 4014 and the computing time is 43 seconds.

**Comment by bernhard-da** [comment]
@olangsrud thx for the tests; i will have a look and try to optimize the computation of required additional constraints;  in this first approach, i compare each constraint (which is slow) with each other and look if exactly two cells overlap excluding common cells which is of course slow; i keep looking for something quicker, though;

as for the "no cells left" error: in this version i was quite restrictive and will not change z-cells to something else either; but this is something that i will need to investigate further.

add gauss-suppression: i just had a quick view; the required input does look quite simular to the full constraint matrix; perhaps it won't be too difficult to reshape it that it fits your required input.
